Source/WebCore/accessibility/AccessibilityNodeObject.cpp

```diff
@@ -3924,6 +3927,194 @@ String AccessibilityNodeObject::text() const
     return element ? element->innerText() : String();
 }
 
+bool AccessibilityNodeObject::isBlockFlow() const
+{
+    return is<RenderBlockFlow>(renderer());
+}
+
+static bool hasEnclosingInputElement(Node* node)
+{
+    return node && is<HTMLInputElement>(node->shadowHost());
+}
+
+static bool hasStitchBreakingRole(Element& element)
+{
+    return hasAnyRole(element, {
+        // Cell roles
+        "gridcell"_s, "cell"_s, "columnheader"_s, "rowheader"_s
+        // Miscellaneous roles
+        "suggestion"_s, "insertion"_s, "deletion"_s
+    });
+}
+
+static bool hasStitchBreakingTag(Element& element)
+{
+    switch (element.elementName()) {
+    case ElementName::HTML_ins:
+    case ElementName::HTML_del:
+        return true;
+    default:
+        return false;
+    }
+}
+
+static bool isStitchBreakingElement(Element& element)
+{
+    return is<HTMLTableCellElement>(element)
+        || is<HTMLLabelElement>(element)
+        || hasStitchBreakingRole(element)
+        || hasStitchBreakingTag(element);
+}
+
+struct StitchingContext {
+    RefPtr<ContainerNode> lastStitchBreakingAncestor { nullptr };
+};
+
+static bool shouldStopStitchingAt(const RenderObject& renderer, StitchingContext& context)
+{
+    if (renderer.style().insideLink() != InsideLink::NotInside) {
+        // Stop stitching when encountering a link.
+        return true;
+    }
+
+    if (renderer.parent() && (renderer.parent()->isBeforeOrAfterContent() || renderer.parent()->isFirstLetter())) {
+        // Stitching generated content will cause incorrect behavior because
+        // some of our code that handles stitched text (e.g. stringValue) assumes
+        // the presence of a Node. For now, stop stitching at generated content.
+        // Ideally we remove this restriction in the future.
+        return true;
+    }
+
+    if (hasEnclosingInputElement(renderer.node())) {
```

Comment:
What about textarea and contenteditable? I feel like it'd be safer to not stitch contenteditable, but I'm not sure. We should test with editing a complex email. That would definitely be a good stress test if things were being constantly stitched and unstitched as you type.

Reply:
Filed issue for us to follow-up on this question.